aten/src/ATen/quantized/Quantizer.cpp

```diff
@@ -479,6 +480,43 @@ QTensorImpl* get_qtensorimpl(const Tensor& self) {
   return static_cast<QTensorImpl*>(self.unsafeGetTensorImpl());
 }
 
+// QNNPACK can access up to 8 bytes beyond the beginning of the tensor's storage
+// boundary which does trigger ASAN, and can result in a segfault if the memory
+// falls on an umapped page or one with no read permission.  Here we define a
+// custom allocator that allocates the extra storage required to keep this
+// behavior safe.
+struct QNNPACKAllocator final : at::Allocator {
+public:
+  virtual ~QNNPACKAllocator() override = default;
+
+  virtual at::DataPtr allocate(size_t nbytes) const override {
+    Cast memory{c10::alloc_cpu(kGuard + nbytes)};
+    memory.as_byte_ptr += kGuard;
```

Comment:
Why do we need to do this here and below (in the deleter)? `memory.as_byte_ptr` doesn't seem to be used outside this struct.

Reply:
This is part of a union.  The variables memory.as_byte_ptr and memory.as_void_ptr occupy the same location in memory, hence modifying one affects the other.  Pointer arithmetic is not allowed on void pointers hence the need for this cast.